### PR TITLE
feature: drift & print strategy config

### PR DESCRIPTION
**Pull request by zenixls2**
also fix `exits` printing
re-implement `CanInt` which doesn't exist in go1.17
call TruncateQuantity on the SubmitOrders and new Trades in backtest

### CODE REVIEW

File: pkg/bbgo/exit.go [hunk 1]
```diff
@@ -35,6 +37,35 @@ type ExitMethod struct {
 	TrailingStop              *TrailingStop2             `json:"trailingStop"`
 }
 
+func (e ExitMethod) String() string {
```
Comment (c9s) [inline pkg/bbgo/exit.go:40]: since it's String(), maybe use YAML format for more readability ?

File: pkg/strategy/drift/output.go [hunk 2]
```diff
@@ -26,6 +22,16 @@ func (a jsonArr) Len() int           { return len(a) }
 func (a jsonArr) Less(i, j int) bool { return a[i].key < a[j].key }
 func (a jsonArr) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 
+func canInt(v reflect.Value) bool {
```
Comment (c9s) [inline pkg/strategy/drift/output.go:25]: Rename to `isInt` and move it to the pkg/dynamic package?

File: pkg/strategy/output.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+func DefaultWhiteList() []string {
+	return []string{"Window", "Interval", "Symbol"}
+}
+
+// @param s: strategy object
+// @param f: io.Writer used for writing the config dump
+// @param style: pretty print table style. Use DefaultStyle() to get default one.
+// @param withColor: whether to print with color
+// @param whiteLists: fields to be printed out from embedded struct (1st layer only)
+func PrintConfig(s interface{}, f io.Writer, style *table.Style, withColor bool, whiteLists ...string) {
```
Comment (c9s) [inline pkg/strategy/output.go:54]: move to pkg/dynamic?

File: pkg/strategy/drift/output.go (hunk 2 above)
Comment (zenixls2) [reply pkg/strategy/drift/output.go:25]: canInt maps to https://pkg.go.dev/reflect#Value.CanInt

File: pkg/bbgo/exit.go (hunk 1 above)
Comment (zenixls2) [reply pkg/bbgo/exit.go:40]: using YAML format will occupy more spaces/lines than json.  Notice that the final output will be inside a column of a table.
also inside ExitMethod, we only define json tags, so YAML format won't work naively.

File: pkg/strategy/output.go (hunk 3 above)
Comment (zenixls2) [reply pkg/strategy/output.go:54]: there'll be a import cycle if I put this in `pkg/dynamic`.

File: pkg/backtest/matching.go [hunk 4]
```diff
@@ -147,6 +147,11 @@ func (m *SimplePriceMatching) PlaceOrder(o types.SubmitOrder) (*types.Order, *ty
 
 	case types.OrderTypeLimit, types.OrderTypeStopLimit, types.OrderTypeLimitMaker:
 		o.Price = m.Market.TruncatePrice(o.Price)
+		// if o.Price makes the limit order a taker, use the latest price as the order price
+		// otherwise we'll endup locking more assets than needed
+		if isLimitTakerOrder(o, m.LastPrice) {
+			o.Price = m.LastPrice
```
Comment (c9s) [inline pkg/backtest/matching.go:153]: maybe you can assign to AveragePrice? the price of the limit order can not be changed.

File: pkg/backtest/matching_test.go [hunk 5]
```diff
@@ -518,7 +518,8 @@ func TestSimplePriceMatching_LimitTakerOrder(t *testing.T) {
 	closedOrder, trade, err = engine.PlaceOrder(newLimitOrder("BTCUSDT", types.SideTypeSell, 19000.0, 1.0))
 	assert.NoError(t, err)
 	if assert.NotNil(t, closedOrder) {
-		assert.Equal(t, "19000", closedOrder.Price.String())
```
Comment (c9s) [inline pkg/backtest/matching_test.go:517]: we need to ensure that the price of the limit order is not changed.

File: pkg/strategy/drift/output.go (hunk 2 above)
Comment (c9s) [reply pkg/strategy/drift/output.go:25]: got it

**Review by c9s — APPROVED** [review]

**Pull request merged**